File: IntersectionTwoLists.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

public class IntersectionTwoLists {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        //if(headA == null || headB == null) return null;

        int len1 = 1;
        int len2 = 1;

        ListNode curr1 = headA;
        ListNode curr2 = headB;

        //length of first one
        while(curr1 != null){
            curr1 = curr1.next;
            len1++;
        }

        //length of second one
        while(curr2 != null){
            curr2 = curr2.next;
            len2++;
        }


        ListNode fast;
        ListNode slow;

        if(len1 > len2){
            fast = headA;
            slow = headB;
        }else{
            fast = headB;
            slow = headA;
        }
        //set lag for slow pointer
        int diff = Math.abs(len1 - len2);
        int count = 0;

        //go until same spot
        while(fast != slow){
            if(count < diff){
                fast = fast.next;
                count++;
            }else{
                fast = fast.next;
                slow = slow.next;
            }
        }

        return fast;

    }
}

File: intersection_two_lists.py
# // Time Complexity : O(n+m) where n,m stand for length of the two lists 
# // Space Complexity : O(1) 
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : none

# // Your code here along with comments explaining your approach 

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        m = n = 0
        p1 = headA
        p2 = headB
        
        while p1:
            m += 1
            p1 = p1.next
    
        while p2:
            n += 1
            p2 = p2.next
        
        p1 = headA
        p2 = headB
        diff = abs(m-n)
        if m > n:
            while diff:
                p1 = p1.next
                diff -= 1
        elif n > m:
            while diff:
                p2 = p2.next
                diff -= 1
        
        while p1 and p2 and p1 != p2:
            
            p1 = p1.next
            p2 = p2.next
            
        if p1 and p2: return p1
        
        return None